toolchain/sem_ir/name_scope_test.cpp

```diff
@@ -0,0 +1,261 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/sem_ir/name_scope.h"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+namespace Carbon::SemIR {
+namespace {
+
+using testing::ElementsAre;
+using testing::Pair;
+
+TEST(NameScope, Empty) {
+  int id = 1;
+
+  InstId scope_inst_id(id++);
+  NameId scope_name_id(id++);
+  NameScopeId parent_scope_id(id++);
+  NameScope name_scope(scope_inst_id, scope_name_id, parent_scope_id);
+
+  EXPECT_THAT(name_scope.entries(), ElementsAre());
+  EXPECT_THAT(name_scope.extended_scopes(), ElementsAre());
+  EXPECT_EQ(name_scope.inst_id(), scope_inst_id);
+  EXPECT_EQ(name_scope.name_id(), scope_name_id);
+  EXPECT_EQ(name_scope.parent_scope_id(), parent_scope_id);
+  EXPECT_FALSE(name_scope.has_error());
+  EXPECT_FALSE(name_scope.is_closed_import());
+  EXPECT_FALSE(name_scope.is_imported_package());
+  EXPECT_THAT(name_scope.import_ir_scopes(), ElementsAre());
+}
+
+TEST(NameScope, Lookup) {
+  int id = 1;
+
+  InstId scope_inst_id(id++);
+  NameId scope_name_id(id++);
+  NameScopeId parent_scope_id(id++);
+  NameScope name_scope(scope_inst_id, scope_name_id, parent_scope_id);
+
+  NameScope::Entry entry1{.name_id = NameId(id++),
+                          .inst_id = InstId(id++),
+                          .access_kind = AccessKind::Public};
+  name_scope.AddRequired(entry1);
+
+  NameScope::Entry entry2{.name_id = NameId(id++),
+                          .inst_id = InstId(id++),
+                          .access_kind = AccessKind::Protected};
+  name_scope.AddRequired(entry2);
+
+  NameScope::Entry entry3{.name_id = NameId(id++),
+                          .inst_id = InstId(id++),
+                          .access_kind = AccessKind::Private};
+  name_scope.AddRequired(entry3);
+
+  auto lookup = name_scope.Lookup(entry1.name_id);
+  ASSERT_NE(lookup, std::nullopt);
+  EXPECT_EQ(StructReflection::AsTuple(
```

Comment:
I've defined a matcher for that.
Using testing::Optional is tricky because I need to call GetEntry with an EntryId so it requires to verify that this isn't nullopt first.